File: packages/xyd-opensdk-java/src/project.ts
import { type NamedType, type OpensdkSpecJson, type ResolvedSdkBehavior, sdkBehavior } from '@xyd-js/opensdk-core';
import type { EmitterContext } from '@xyd-js/opensdk-framework';

import { javaPackageName } from './naming';
import type { OpensdkJavaOptions } from './types';

/** Per-file emit context: package layout, symbol table, auth, resolved behavior. */
export interface JavaCtx {
  /** The leaf package segment, e.g. "petstore". */
  pkg: string;
  /** The base package prefix, e.g. "com.example". */
  basePackage: string;
  /** The full package, e.g. "com.example.petstore". */
  fullPackage: string;
  /** The source directory prefix every .java file nests under (Maven layout). */
  srcDir: string;
  types: Map<string, NamedType>;
  behavior: ResolvedSdkBehavior;
  baseURL: string;
  /** The env var the client seeds its credential from, or null when unauthenticated. */
  envVar: string | null;
  /** The normalized auth scheme kind (bearer | apiKey-header | apiKey-query | ...), or null. */
  authKind: string | null;
  /** The header/query name for apiKey auth, or null. */
  authName: string | null;
  /** The SDK identifier baked into the User-Agent, e.g. "petstore-java/1.2.0". */
  userAgent: string;
}

export function resolveJavaOptions(spec: OpensdkSpecJson, ctx: EmitterContext): JavaCtx {
  const options = ctx.emitterOptions as OpensdkJavaOptions;
  const pkg = options.packageName ?? javaPackageName(spec.info.title);
  const basePackage = options.basePackage ?? 'com.example';
  const fullPackage = `${basePackage}.${pkg}`;
  const security = spec.security?.[0];
  const behavior = sdkBehavior(spec);
  const userAgent = behavior.userAgent.sdkIdentifierTemplate
    .split('{package}')
    .join(pkg)
    .split('{language}')
    .join('java')
    .split('{version}')
    .join(spec.info.version || '0.0.0');
  return {
    pkg,
    basePackage,
    fullPackage,
    srcDir: `src/main/java/${fullPackage.split('.').join('/')}/`,
    types: ctx.types as Map<string, NamedType>,
    behavior,
    baseURL: options.baseURL ?? spec.servers?.[0] ?? '',
    envVar: spec.security?.find((s) => s.envVar)?.envVar ?? null,
    authKind: security?.kind ?? null,
    authName: security?.name ?? null,
    userAgent,
  };
}

/**
 * `pom.xml` — the build manifest. Dependency-free (the runtime is stdlib-only),
 * so it declares only the compiler release. User-owned scaffold: never clobber
 * an existing pom on regen (mirrors the Go/Python go.mod/pyproject writeMode).
 * The generated source tree is equally compilable by a bare
 * `javac $(find . -name '*.java')`.
 */
/** Minimal XML text escape for interpolated identity values. */
function xmlEscape(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

export function pomXml(ctx: JavaCtx, spec: OpensdkSpecJson): string {
  const info = spec.info;
  // Publish identity (name/url/licenses/developers/scm) from spec.info — the
  // OpenAPI `info` or an sdk.json `publish` override; enough metadata for
  // `mvn deploy` (Central signing is a separate release concern).
  const meta: string[] = [`  <name>${xmlEscape(info.title)}</name>`];
  if (info.homepage) meta.push(`  <url>${xmlEscape(info.homepage)}</url>`);
  if (info.license?.identifier) {
    meta.push(
      '  <licenses>',
      `    <license><name>${xmlEscape(info.license.identifier)}</name></license>`,
      '  </licenses>',
    );
  }
  if (info.contact?.name) {
    meta.push(
      '  <developers>',
      `    <developer><name>${xmlEscape(info.contact.name)}</name></developer>`,
      '  </developers>',
    );
  }
  if (info.repository) meta.push('  <scm>', `    <url>${xmlEscape(info.repository)}</url>`, '  </scm>');
  return `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>${ctx.basePackage}</groupId>
  <artifactId>${ctx.pkg}-java</artifactId>
  <version>${info.version || '0.0.0'}</version>
  <packaging>jar</packaging>
${meta.join('\n')}

  <properties>
    <maven.compiler.release>11</maven.compiler.release>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
  </properties>
</project>
`;
}

File: crates/xyd_opensdk_java/src/project.rs
//! project.ts — the per-file JavaCtx (resolveJavaOptions) + pom.xml. Emitter
//! options default to empty (the back-compat `opensdkJava(spec)` path).

use serde_json::Value;

use crate::behavior::resolve_behavior;
use crate::ir::{str_field, Types};
use crate::jsrt::java_package_name;

pub struct JavaCtx {
    pub pkg: String,
    pub base_package: String,
    pub full_package: String,
    pub src_dir: String,
    pub types: Types,
    pub base_url: String,
    pub env_var: Option<String>,
    /// The resolved `ResolvedSdkBehavior` block (defaults ⊕ `spec.sdk`); drives
    /// the vendored runtime's retry/timeout/error/user-agent policy.
    pub behavior: Value,
    /// The normalized auth scheme kind (bearer | apiKey-header | ...), or None.
    pub auth_kind: Option<String>,
    /// The header/query name for apiKey auth, or None.
    pub auth_name: Option<String>,
    /// The SDK identifier baked into the User-Agent, e.g. "petstore-java/1.2.0".
    pub user_agent: String,
}

pub fn resolve_java_options(spec: &Value, types: Types) -> JavaCtx {
    let title = str_field(spec.get("info").unwrap_or(&Value::Null), "title").unwrap_or("");
    let pkg = java_package_name(title);
    let base_package = "com.example".to_string();
    let full_package = format!("{base_package}.{pkg}");
    let src_dir = format!(
        "src/main/java/{}/",
        full_package.split('.').collect::<Vec<_>>().join("/")
    );

    let base_url = spec
        .get("servers")
        .and_then(|s| s.as_array())
        .and_then(|a| a.first())
        .and_then(|s| s.as_str())
        .unwrap_or("")
        .to_string();
    let env_var = spec
        .get("security")
        .and_then(|s| s.as_array())
        .and_then(|arr| arr.iter().find(|s| s.get("envVar").is_some()))
        .and_then(|s| str_field(s, "envVar"))
        .map(|s| s.to_string());

    let security = spec
        .get("security")
        .and_then(|s| s.as_array())
        .and_then(|a| a.first());
    let auth_kind = security
        .and_then(|s| str_field(s, "kind"))
        .map(|s| s.to_string());
    let auth_name = security
        .and_then(|s| str_field(s, "name"))
        .map(|s| s.to_string());

    let behavior = resolve_behavior(spec);
    let version = str_field(spec.get("info").unwrap_or(&Value::Null), "version")
        .filter(|v| !v.is_empty())
        .unwrap_or("0.0.0");
    let template = behavior
        .get("userAgent")
        .and_then(|u| u.get("sdkIdentifierTemplate"))
        .and_then(|t| t.as_str())
        .unwrap_or("{package}-{language}/{version}");
    let user_agent = template
        .replace("{package}", &pkg)
        .replace("{language}", "java")
        .replace("{version}", version);

    JavaCtx {
        pkg,
        base_package,
        full_package,
        src_dir,
        types,
        base_url,
        env_var,
        behavior,
        auth_kind,
        auth_name,
        user_agent,
    }
}

fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

/// pom.xml — the (dependency-free) build manifest.
pub fn pom_xml(ctx: &JavaCtx, spec: &Value) -> String {
    let info = spec.get("info").cloned().unwrap_or(Value::Null);
    let title = str_field(&info, "title").unwrap_or("");
    let version = str_field(&info, "version")
        .filter(|v| !v.is_empty())
        .unwrap_or("0.0.0");

    let mut meta: Vec<String> = vec![format!("  <name>{}</name>", xml_escape(title))];
    if let Some(h) = str_field(&info, "homepage") {
        meta.push(format!("  <url>{}</url>", xml_escape(h)));
    }
    if let Some(lic) = info.get("license").and_then(|l| str_field(l, "identifier")) {
        meta.push("  <licenses>".to_string());
        meta.push(format!(
            "    <license><name>{}</name></license>",
            xml_escape(lic)
        ));
        meta.push("  </licenses>".to_string());
    }
    if let Some(cn) = info.get("contact").and_then(|c| str_field(c, "name")) {
        meta.push("  <developers>".to_string());
        meta.push(format!(
            "    <developer><name>{}</name></developer>",
            xml_escape(cn)
        ));
        meta.push("  </developers>".to_string());
    }
    if let Some(r) = str_field(&info, "repository") {
        meta.push("  <scm>".to_string());
        meta.push(format!("    <url>{}</url>", xml_escape(r)));
        meta.push("  </scm>".to_string());
    }

    format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>{base}</groupId>
  <artifactId>{pkg}-java</artifactId>
  <version>{version}</version>
  <packaging>jar</packaging>
{meta}

  <properties>
    <maven.compiler.release>11</maven.compiler.release>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
  </properties>
</project>
"#,
        base = ctx.base_package,
        pkg = ctx.pkg,
        version = version,
        meta = meta.join("\n"),
    )
}
